src/hotspot/share/nmt/indexedFreeListAllocator.hpp

```diff
@@ -0,0 +1,149 @@
+#include "memory/allocation.hpp"
+#include "memory/arena.hpp"
+#include "utilities/growableArray.hpp"
+#include "runtime/os.hpp"
+
+// A free list, growth only, allocator for a specific type E.
+// The allocator returns 'pointers' of 4-bytes in size, allowing for
+// memory savings if a pointer-heavy self-referential structure is used.
+// It is "indexed" as a reference is base + index * sizeof(E).
+// It never returns any memory to the system.
+template<typename E, MEMFLAGS flag>
+class IndexedFreeListAllocator {
+public:
+  // Make the index opaque.
+  struct I {
+    int32_t idx;
+    bool operator !=(I other) {
+      return idx != other.idx;
+    }
+    bool operator==(I other) {
+      return idx == other.idx;
+    }
+  };
+  static constexpr const I nil = I{-1};
+  // A free list allocator element is either a link to the next free space
+  // Or an actual element.
+  union BackingElement {
+    I link;
+    E e;
+    BackingElement() {
+      this->link = nil;
+    }
+    BackingElement(I link) {
+      this->link = link;
+    }
+    BackingElement(E& e) {
+      this->e = e;
+    }
+  };
+  GrowableArrayCHeap<BackingElement, flag> backing_storage;
+  I free_start;
+
+  IndexedFreeListAllocator()
+  : backing_storage(8),
+  free_start(I{0}) {}
+
+  template<typename... Args>
+  I allocate(Args... args) {
+    int32_t i = free_start.idx;
+    backing_storage.at_grow(i);
+    BackingElement& be = backing_storage.at(i);
+    if (be.link == nil) {
+      // Must be at end, simply increment
+      free_start.idx += 1;
+    } else {
+      // Follow the link to the next free element
+      free_start = be.link;
+    }
+    ::new (&be) E(args...);
+    return I{i};
+  }
+
+  void free(I i) {
+    BackingElement& be_freed = backing_storage.at(i.idx);
+    be_freed.link = free_start;
+    free_start = i;
+  }
+
+  E& operator[](I i) {
+    return backing_storage.at(i.idx).e;
+  }
+
+  E& translate(I i) {
+    return backing_storage.at(i.idx).e;
+  }
+};
+
```

Comment:
What are these allocators for?

Reply:
So you can easily swap out your allocator if you have a bug and want to ensure the bug isn't in your allocator. Also just completeness, and necessary for the performance tests. We can remove them, but I think they're nice to have.